Adding new documentation for the local mode to run the AI-powered transpiler passes (#2273)
Qiskit/documentation: docs/guides/ai-transpiler-passes.mdx

# AI transpiler passes
​

> The AI-powered transpiler passes are experimental passes that work as a drop-in replacement of "traditional" Qiskit passes for some transpiling tasks. They often produce better results than existing heuristic algorithms (such as lower depth and CNOT count), but are also much faster than optimization algorithms such as Boolean satisfiability solvers. The AI transpiler passes can run on your local environment or on the cloud using the Qiskit Transpiler Service if you are part of the IBM Quantum&trade; Premium Plan.

Reviewer: Can all AI transpiler passes run locally, or just the routing pass? Do they have the same exact capabilities?

Reviewer: I saw `experimental` was taken out in the intro. Should it be taken out here as well?

Author: Yes, all the transpiler passes can run on local except the new one for PauliNetworks https://github.com/Qiskit/documentation/pull/2263

Author: Ok, I'll remove the experimental word. Thanks

Author: Fixed in https://github.com/Qiskit/documentation/pull/2273/commits/ab96f534a495708347f73a803bbd3c8ed2c8b073

> The AI-powered transpiler passes are experimental passes that work as a drop-in replacement of "traditional" Qiskit passes for some transpiling tasks. They often produce better results than existing heuristic algorithms (such as lower depth and CNOT count), but are also much faster than optimization algorithms such as Boolean satisfiability solvers. The AI transpiler passes can run on your local environment or on the cloud using the Qiskit Transpiler Service if you are part of the IBM Quantum&trade; Premium Plan users.

Reviewer 2: ```suggestion
The AI-powered transpiler passes are experimental passes that work as a drop-in replacement of "traditional" Qiskit passes for some transpiling tasks. They often produce better results than existing heuristic algorithms (such as lower depth and CNOT count), but are also much faster than optimization algorithms such as Boolean satisfiability solvers. The AI transpiler passes can run on your local environment or on the cloud using the Qiskit Transpiler Service if you are part of the IBM Quantum&trade; Premium Plan.
```

 - `AILinearFunctionSynthesis`: Linear function circuit synthesis
 - `AIPermutationSynthesis`: Permutation circuit synthesis

> To use the AI transpiler passes, install the `qiskit-ibm-transpiler` package (see instructions [here](./qiskit-transpiler-service#install-transpiler-service)).

Reviewer 2: ```suggestion
To use the AI transpiler passes, first [install the `qiskit-ibm-transpiler` package.](./qiskit-transpiler-service#install-transpiler-service))
```

> ## Running the AI transpiler passes on local or cloud

Reviewer 2: ```suggestion
## Run the AI transpiler passes locally or on cloud
```

> To use the AI transpiler passes, first [install the `qiskit-ibm-transpiler` package.](./qiskit-transpiler-service#install-transpiler-service). Check out the [qiskit-ibm-transpiler API documentation](https://docs.quantum.ibm.com/api/qiskit-ibm-transpiler) to get more information about the different options available.

Reviewer 3: ```suggestion
To use the AI transpiler passes, first [install the `qiskit-ibm-transpiler` package](/guides/qiskit-transpiler-service#install-transpiler-service). Visit the [qiskit-ibm-transpiler API documentation](/api/qiskit-ibm-transpiler) to get more information about the different options available.
```

> If you want to use the AI-powered transpiler passes in your local environment for free you will need install the `qiskit-ibm-transpiler` with some extra dendencies as follows:

Reviewer 2: ```suggestion
If you want to use the AI-powered transpiler passes in your local environment for free you need install `qiskit-ibm-transpiler` with some extra dependencies as follows:
```

> If you want to use the AI-powered transpiler passes in your local environment for free you need install `qiskit-ibm-transpiler` with some extra dependencies as follows:

Reviewer 3: ```suggestion
If you want to use the AI-powered transpiler passes in your local environment for free, install `qiskit-ibm-transpiler` with some extra dependencies as follows:
```

```sh
pip install qiskit-ibm-transpiler["ai-local-mode"]
```

> Without these extra dependencies, the AI-powered transpiler passes will run on cloud through the Qiskit Transpiler Service (available only for IBM Quantum&trade; Premium Plan users). After installing the extra dependencies, the default mode to run the AI-powered transpiler passes will be using your local machine.

Reviewer 2: ```suggestion
Without these extra dependencies, the AI-powered transpiler passes run on cloud through the Qiskit Transpiler Service (available only for IBM Quantum&trade; Premium Plan users). After installing the extra dependencies, the default mode to run the AI-powered transpiler passes is to use your local machine.
```

- `keep`: This respects the layout set by the previous transpiler passes (or uses the trivial layout if not set). It is typically only used when the circuit must be run on specific qubits of the device. It often produces worse results because it has less room for optimization.
- `improve`: This uses the layout set by the previous transpiler passes as a starting point. It is useful when you have a good initial guess for the layout; for example, for circuits that are built in a way that approximately follows the device's coupling map. It is also useful if you want to try other specific layout passes combined with the `AIRouting` pass.
- `optimize`: This is the default mode. It works best for general circuits where you might not have good layout guesses. This mode ignores previous layout selections.

> - `local_mode`: This flag determines where the AIRouting pass runs, if `False`, AIRouting runs on cloud through the Qiskit Transpiler Service. If `True`, the package tries to run on your local environment with a fallback to cloud mode if the required the extra dependencies are not found.

Reviewer: ```suggestion
- `local_mode`: This flag determines where the `AIRouting` pass runs. If `False`, `AIRouting` runs remotely through the Qiskit Transpiler Service. If `True`, the package tries to run the pass in your local environment with a fallback to cloud mode if the required dependencies are not found.
```

> - `local_mode`: This flag determines where your AIRouting pass will run, if `False`, AIRouting will run on cloud through the Qiskit Transpiler Service. If `True` the package will try to run on your local environment with a fallback to the cloud mode if the required the extra dependencies needed to run the AI-powered transpiler passes on local are not found.

Reviewer 2: ```suggestion
- `local_mode`: This flag determines where the AIRouting pass runs, if `False`, AIRouting runs on cloud through the Qiskit Transpiler Service. If `True`, the package tries to run on your local environment with a fallback to cloud mode if the required the extra dependencies are not found.
```

These custom collection passes limit the sizes of the collected sub-circuits so they are supported by the AI-powered synthesis passes. Therefore, it is recommended to use them after the routing passes and before the synthesis passes for a better overall optimization.

## Hybrid Qiskit-AI circuit transpilation

> The `qiskit-ibm-transpiler` offers the option to configure an hybrid pass manager combining best of Qiskit and the AI-powered transpiler passes, so you can run a whole transpilation on your own. This featres behaves similarly to the Qiskit `generate_pass_manager` method. A typical way to use this `generate_ai_pass_manager` is as follows:

Reviewer 2: ```suggestion
The `qiskit-ibm-transpiler` offers the option to configure a hybrid pass manager that combines the best of Qiskit and the AI-powered transpiler passes, so you can run a whole transpilation on your own. This feature behaves similarly to the Qiskit `generate_pass_manager` method. A typical way to use `generate_ai_pass_manager` is as follows:
```

```python
    from qiskit_ibm_transpiler.ai import generate_ai_pass_manager
    from qiskit.circuit.library import EfficientSU2
    from qiskit_ibm_runtime import QiskitRuntimeService

    backend = QiskitRuntimeService().backend("ibm_fez")
    fez_coupling_map = backend.coupling_map

    su2_circuit = EfficientSU2(101, entanglement="circular", reps=1).decompose()

    ai_transpiler_pass_manager = generate_ai_pass_manager(,
        coupling_map=fez_coupling_map,
        ai_optimization_level=3,
        optimization_level=3,
        ai_layout_mode=optimize,
    )

    ai_su2_transpiled_circuit = ai_transpiler_pass_manager.run(su2_circuit),
```

> Here, the `coupling_map` determines which coupling map needs to be used for the transpilation. The `ai_optimization_level` refers to the level of optimization (1-3) to get using the AI components of the PassManager, while the `optimization_level` refers to the optimization to get from the Qiskit heuristic algorithms. The `ai_layout_mode` refers to how the AI routing part of the PassManager will handle the layout. Refer to the #AI-routing-pass section to review the different configuration options for this parameter.

Reviewer 2: ```suggestion
The following options are used:
* `coupling_map` - Specifies which coupling map to use for the transpilation. 
* `ai_optimization_level` - Specifies the level of optimization (1-3) to use for the AI components of the PassManager.
* `optimization_level` - Specifies the optimization to use for the Qiskit heuristic algorithms. 
* `ai_layout_mode` - Specifies how the AI routing part of the PassManager handles the layout. 

Refer to the #AI-routing-pass section to review the configuration options for this parameter.
```

```python
    from qiskit_ibm_transpiler.ai import generate_ai_pass_manager
    from qiskit.circuit.library import EfficientSU2
    from qiskit_ibm_runtime import QiskitRuntimeService

    backend = QiskitRuntimeService().backend("ibm_fez")
    fez_coupling_map = backend.coupling_map

    su2_circuit = EfficientSU2(101, entanglement="circular", reps=1).decompose()

    ai_transpiler_pass_manager = generate_ai_pass_manager(,
        coupling_map=fez_coupling_map,
        ai_optimization_level=3,
        optimization_level=3,
        ai_layout_mode="optimize",
    )

    ai_su2_transpiled_circuit = ai_transpiler_pass_manager.run(su2_circuit),
```

The following options are used in this example:
* `coupling_map` - Specifies which coupling map to use for the transpilation.
* `ai_optimization_level` - Specifies the level of optimization (1-3) to use for the AI components of the PassManager.
* `optimization_level` - Specifies how much optimization to perform on the circuit for the heuristic components of the PassManager.

> * `ai_layout_mode` - Specifies how the AI routing part of the PassManager handles the layout. Refer to the #AI-routing-pass section to review the configuration options for this `ai_layout_mode` parameter.

Reviewer 3: ```suggestion
* `ai_layout_mode` - Specifies how the AI routing part of the PassManager handles the layout. Refer to the [AI routing pass](#ai-routing-pass) section to review the configuration options for this `ai_layout_mode` parameter.
```

These custom collection passes limit the sizes of the collected sub-circuits so they are supported by the AI-powered synthesis passes. Therefore, it is recommended to use them after the routing passes and before the synthesis passes for a better overall optimization.

> ## Hybrid Qiskit-AI circuit transpilation

Reviewer: ```suggestion
## Hybrid heuristic-AI circuit transpilation
```

I think calling it Qiskit might be a bit confusing since we use "Qiskit" also as a brand in addition to the SDK itself.

> The `qiskit-ibm-transpiler` offers the option to configure a hybrid pass manager that combines the best of Qiskit and the AI-powered transpiler passes, so you can run a whole transpilation on your own. This feature behaves similarly to the Qiskit `generate_pass_manager` method. A typical way to use `generate_ai_pass_manager` is as follows:

Reviewer: ```suggestion
The `qiskit-ibm-transpiler` allows you to configure a hybrid pass manager that combines the best of Qiskit's heuristic and the AI-powered transpiler passes. This feature behaves similarly to the Qiskit `generate_pass_manager` method. A typical way to use `generate_ai_pass_manager` is as follows:
```

I took out `so you can run a whole transpilation on your own` because I don't understand what that means.

Author: Ok, I do agree it's more clear now. Thanks

```python
    from qiskit_ibm_transpiler.ai import generate_ai_pass_manager
    from qiskit.circuit.library import EfficientSU2
    from qiskit_ibm_runtime import QiskitRuntimeService

    backend = QiskitRuntimeService().backend("ibm_fez")
    fez_coupling_map = backend.coupling_map

    su2_circuit = EfficientSU2(101, entanglement="circular", reps=1).decompose()

    ai_transpiler_pass_manager = generate_ai_pass_manager(,
        coupling_map=fez_coupling_map,
        ai_optimization_level=3,
        optimization_level=3,

>         ai_layout_mode=optimize,

Reviewer: Is `optimize` supposed to be a string literal or a variable? If the latter, it's not defined. Also, can I set `local_mode` here?

    )

    ai_su2_transpiled_circuit = ai_transpiler_pass_manager.run(su2_circuit),
```

> The following options are used:

Reviewer: Does this mean these are the options used in the code example, or these options are available? Also should link to API reference.

Author: Added the link to the API reference

* `coupling_map` - Specifies which coupling map to use for the transpilation. 

> * `ai_optimization_level` - Specifies the level of optimization (1-3) to use for the AI components of the PassManager.

Reviewer: Is `ai_optimization_level` also a new option for `TranspilerService`?

Author: No, it's only available for this passmanager

> * `optimization_level` - Specifies the optimization to use for the Qiskit heuristic algorithms. 

Reviewer: ```suggestion
* `optimization_level` - Specifies how much optimization to perform on the circuit for the heuristic components of the PassManager.
```

* `ai_layout_mode` - Specifies how the AI routing part of the PassManager handles the layout. 

> Refer to the #AI-routing-pass section to review the configuration options for this parameter.

Reviewer: What parameter?

Author: It has a line break that shouldn't be there. Fixed in https://github.com/Qiskit/documentation/pull/2273/commits/aa8575f5eaad17a6bbe6b6a913fb3f563c925ee0
